Sources/NIO/Bootstrap.swift

```diff
@@ -966,4 +1133,138 @@ public final class NIOPipeBootstrap {
             }
         }
     }
+    
+    /// A channel option which can be applied to pipe bootstrap using shorthand notation.
+    /// - See: NIOPipeBootstrap.channelOptions(_ options: [Option])
+    public struct Option {
+        private let data: ShorthandChannelOption
+        
+        private init(_ data: ShorthandChannelOption) {
+            self.data = data
+        }
+        
+        /// Apply the contained option to the supplied NIOPipeBootstrap
+        /// - Parameter to: bootstrap to apply this option to.
+        @usableFromInline
+        func applyOption(to bootstrap: NIOPipeBootstrap) {
+            data.applyOption(to: bootstrap)
+        }
+        
+        fileprivate enum ShorthandChannelOption {
+            case disableAutoRead
+            case allowRemoteHalfClosure(Bool)
+            
+            func applyOption(to bootstrap: NIOPipeBootstrap) {
+                switch self {
+                case .disableAutoRead:
+                    _ = bootstrap.channelOption(ChannelOptions.autoRead, value: false)
+                case .allowRemoteHalfClosure(let value):
+                    _ = bootstrap.channelOption(ChannelOptions.allowRemoteHalfClosure, value: value)
+                }
+            }
+        }
+    }
+}
+
+// Hashable for the convenience of users.
+extension NIOPipeBootstrap.Option: Hashable {}
+extension NIOPipeBootstrap.Option.ShorthandChannelOption: Hashable {}
+
+/// Approved shorthand datagram channel options.
+extension NIOPipeBootstrap.Option {
+    /// Option to disable autoRead
+    /// - See: ChannelOptions.autoRead
+    public static let disableAutoRead = NIOPipeBootstrap.Option(.disableAutoRead)
+    
+    /// - See: `AllowRemoteHalfClosureOption`.
+    public static let allowRemoteHalfClosure =
+        NIOPipeBootstrap.Option(.allowRemoteHalfClosure(true))
+    
+    /// - See: `AllowRemoteHalfClosureOption`.
+    public static func allowRemoteHalfClosure(_ value: Bool) ->
+        NIOPipeBootstrap.Option {
+        return NIOPipeBootstrap.Option(.allowRemoteHalfClosure(value))
+    }
+}
+
+
+
+// ------------------------
+
+@usableFromInline
+internal protocol NIOTCPOptionAppliable {
+    func applyOption<Option: ChannelOption>(_ option: Option, value: Option.Value) -> Self
+}
+
+extension ServerBootstrap : NIOTCPOptionAppliable {
+    @usableFromInline
+    func applyOption<Option>(_ option: Option, value: Option.Value) -> Self where Option : ChannelOption {
+        return self.childChannelOption(option, value: value)
+    }
+}
+
+extension ClientBootstrap : NIOTCPOptionAppliable {
+    @usableFromInline
+    func applyOption<Option>(_ option: Option, value: Option.Value) -> Self where Option : ChannelOption {
+        return self.channelOption(option, value: value)
+    }
+}
```

Comment:
Weissi asks:  Why does ClientBootstrap etc need a conformance to NIOTCPOptionAppliable but NIOTSConnectionBootstrap does not?

Answer to open the discussion:  

The purpose of this little protocol was to allow pointing at the various ways of converting shorthand options to old channel options - eg childChannelOption on server vs channelOption on client.  I expect it could probably be done with a function type if I was smart enough to work out how to deal with the Self types.

Why does NIOTS not have it - because it has no direct shorthand options currently.  The only way you can currently use shorthand options with NIOTS is via the universal bootstrap.  

If I was to add the shorthand options directly to NIOTS it would need conformance (in which case this protocol needs to be public - maybe it should be anyway).

Reply:
Makes sense. I think `ClientBootstrap` and `NIOTSConnectionBootstrap` should be treated as equal citizens. Either we support convenience on neither (only through uni boot) or on both.

Reply:
I'll make this public and add NIOTS quickly, there's bound to be someone else out there with their own bootstrap which might appreciate it.

Reply:
I have made the protocol public but changed the implementation slightly so that it's more like a closure so the bootstrap API doesn't have to be polluted with the extra method.  Renamed NIOChannelOptionAppliable for general use across the different option types.